trillian/ctfe/handlers.go

```diff
@@ -461,13 +466,35 @@ func addChainInternal(ctx context.Context, li *logInfo, w http.ResponseWriter, r
 	timeMillis := uint64(li.TimeSource.Now().UnixNano() / millisPerNano)
 
 	// Build the MerkleTreeLeaf that gets sent to the backend, and make a trillian.LogLeaf for it.
+	var leaf trillian.LogLeaf
 	merkleLeaf, err := ct.MerkleTreeLeafFromChain(chain, etype, timeMillis)
 	if err != nil {
 		return http.StatusBadRequest, fmt.Errorf("failed to build MerkleTreeLeaf: %s", err)
 	}
-	leaf, err := buildLogLeafForAddChain(li, *merkleLeaf, chain, isPrecert)
-	if err != nil {
-		return http.StatusInternalServerError, fmt.Errorf("failed to build LogLeaf: %s", err)
+	raw := extractRawCerts(chain)
+
+	// If CTFE storage is enabled for issuance chain, add the chain to storage
+	// and cache, and then build log leaf. If Trillian gRPC is enabled for
+	// issuance chain, build the log leaf.
+	if li.issuanceChainService.IsCTFEStorageEnabled() {
+		// TODO: Check how to convert []ct.ASN1Cert to []byte correctly.
+		issuanceChain, err := asn1.Marshal(raw[1:])
+		if err != nil {
+			return http.StatusInternalServerError, fmt.Errorf("failed to marshal issuance chain: %s", err)
+		}
+		hash, err := li.issuanceChainService.Add(ctx, issuanceChain)
+		if err != nil {
+			return http.StatusInternalServerError, fmt.Errorf("failed to add issuance chain into CTFE storage: %s", err)
+		}
+		leaf, err = util.BuildLogLeafWithHash(li.LogPrefix, *merkleLeaf, 0, raw[0], nil, hash, isPrecert)
+		if err != nil {
+			return http.StatusInternalServerError, fmt.Errorf("failed to build LogLeaf: %s", err)
+		}
+	} else {
+		leaf, err = util.BuildLogLeaf(li.LogPrefix, *merkleLeaf, 0, raw[0], raw[1:], isPrecert)
+		if err != nil {
+			return http.StatusInternalServerError, fmt.Errorf("failed to build LogLeaf: %s", err)
+		}
```

Comment:
This pattern is essentially:

```
if x.IsFeatureSupported(){
  return x.UseFeature()
} else {
  return someOtherStrategy()
}
```

I would prefer to see these strategies simply moved into the `IssuanceChainService`. From here, the code would just be `leaf, err := issuanceChainService.BuildLeaf(...)`. The complexity is moved to `IssuanceChainService`, which could do with some refactoring. `IssuanceChainService` is currently a simple implementation with an optional field (`storage`) that fundamentally changes its behaviour, suggesting to me that it should be an interface with two different implementations (a storage backed one, and a non-storage backed one).

You may then be able to remove the `Add` method from `IssuanceChainService`, because it would be an internal detail called from the storage-backed `BuildLeaf`. The noop chain service would not need to call/define add. 

This might seem like a lot of work for no gain, but the separation of concerns avoids bugs. As a data point, calling the following code here would cause a panic as the code currently stands:
```
if !li.issuanceChainService.IsCTFEStorageEnabled() {  li.issuanceChainService.Add(ctx, issuanceChain) }
```

This is because you don't check for nil storage in `Add` at the moment. It's an easy enough fix, but if the concerns were separated into two implementations then this bug wouldn't be possible.

Reply:
This is a really good analysis. The original consideration is to minimise the changes to the existing logic and flow. However, it increases the tech debt with a lower code readability.

Good catch for the panic when `storage` is `nil`. I've thought about that and added a noop implementation for the cache. It is true that someone else may call the `Add` method without checking whether the `storage` is `nil` or not.

Reply:
Calling `BuildLeaf(...)` in `issuanceChainService` seems to be out of the scope of the service's responsibility as the issuance chain is only part of the log leaf.